PEP 715: Disabling bdist_egg distribution uploads on PyPI (#3161)
python/peps: pep-9999.rst

> PEP: 9999

Reviewer: Suggestion: Assign PEP: 6969

Reviewer 2: Could you explain the rationale for this? I'm not sure I understand the reasoning or get the joke, sorry (other than the crude sexual humor, which I'm assuming this is not intended as a reference to, as I don't see how the PR has a clear connection to that—though many readers are likely to assume that).

In general, as the [relevant section of PEP 1 states](https://peps.python.org/pep-0001/#pep-editor-responsibilities-workflow)...

> Assign a PEP number (almost always just the next available number, but sometimes it’s a special/joke number, like 666 or 3141).

...special PEP numbers have reserved for the very rare, "special"-case or "joke" PEPs, of which historically there have only been around a half dozen or so out of over 600 PEPs), which all had a clear, compelling and immediately obvious reason to use the specific number, and the most recent of which was many years ago.

As far as I'm aware, while there's been a few requests for vanity numbers more recently, including by core devs, there hasn't been a compelling reason to deviate from the standard, established pattern in recent times, given it makes the ordering more useful, meaningful, predictable and consistent, and avoids favoritism and potential confusion as to why one PEP gets a "special" out of sequence number while another does not.

Also, this particular request would raise more significant concerns than those before, as it would be not only out of sequence but a 4-digit number in a brand new thousands range, above that of an existing reservation (and below another), which up until now have always carried semantic meaning and been used for special series of PEPs (Python 3k and governance), and which may be needed for others in the future (Yes, we'll likely have 1xxxx PEPs within about 5-10 years at the current rate, but they will be contiguous with the current series, and should last us for at least another few decades assuming this repo survives that long).

Title: Disabling bdist_egg distribution uploads on PyPI
Author: William Woodruff <william@yossarian.net>
PEP-Delegate: Donald Stufft <donald@stufft.io>

> Sponsor: Donald Stufft <donald@stufft.io>

Reviewer 3: Just documenting that I've explicitly OK'd William to use me as a Sponsor.

* ``build`` only supports the ``sdist`` and ``bdist_wheel`` filetypes.

Given the above, this PEP proposes the removal of the ``bdist_egg`` format
under the same justifications presented in :pep:`527`, namely:

1. Egg distributions are of of limited use to the broader ecosystem and
   therefore represent a non-reciprocal maintenance burden;
2. Continued support for Egg distributions represents a bandwidth and storage
   cost to PyPI, as well as to any external mirrors of PyPI.

This PEP *additionally* offers an argument for removal rooted in
standardization and duplication: the Egg format is not standardized by any
PEP or other community standard, and overlaps heavily with its standardized
and well-supported alternative (Wheel).

The ``.egg`` file extension
---------------------------

The ``.egg`` file extension is used exclusively for distributions of the
``bdist_egg`` format. As such, it serves no purpose in a scenario where
PyPI disables new distribution uploads for ``bdist_egg`` distributions.

Removal Process
===============

This PEP proposes a removal process based on :pep:`527`, but with the following
modifications:

> * The deprecation period will be six months, rather than one month.

Reviewer: Suggestion: Stick with one month. There was outreach done in the initial warehouse issue to determine outstanding usecases of egg uploads. At that time eggs were already on the decline for uploads.

Question: Would updated metrics/stats for egg uploads the helpful to motivate moving this to a more accelerated timeline?

Author: Sounds good to me -- I arbitrarily chose 6 months because of the more "aggressive" email policy (don't email all current egg uploaders, only upload if they upload another egg), but IMO 1 month is also suitable 🙂

Author: fb4c3847

Reviewer: Volumes are low enough I'm actually game to do the one-time email with a one month deprecation.

Reviewer: Peak .egg was `2020-05` with `1534` uploaded files.

Here's the last 12 full months for .egg files uploaded:

```
 2022-12 |   531
 2023-01 |   653
 2023-02 |   695
 2023-03 |   570
 2023-04 |   620
 2023-05 |   576
```

Reviewer: all 1022 projects and the number of eggs they've uploaded, in 2022: https://gist.github.com/ewdurbin/d7939458f6205069adeeb18c454c9331

Reviewer 4: I'm in favor of a one-time bulk email + a 1 mo deprecation period with emails per upload.

Author: 92a85aa3

> PEP: 9999

Reviewer 2: Reminder to update this again once all review suggestions are applied

Post-History:
Resolution:

Abstract
========

This PEP recommends deprecating and then disabling new uploads of the
``bdist_egg`` distribution type on PyPI. In a parallel move, this PEP recommends
deprecating and then disabling new uploads of distribution filenames that have
the ``.egg`` suffix.

It does not recommend removing or otherwise affecting any previously
uploaded ``bdist_egg`` distributions.

Rationale
=========

Previous Work
-------------

The groundwork for this proposal was established with :pep:`527`, which
proposed deprecating and eventually removing upload support for a handful
of un(der)used file extensions and distribution types.

In particular, :pep:`527` proposed the removal of the ``bdist_dumb``,
``bdist_rpm``, ``bdist_dmg``, ``bdist_msi``, and ``bdist_wininst`` distribution
types, as well as the ``.tar``, ``.tar.bz2``, ``.tar.xz``, ``.tar.Z``,
``.tgz``, and ``.tbz`` file extensions on distribution filenames.

:pep:`527` was fully enacted with
`PR #7529 <https://github.com/pypi/warehouse/pull/7529>`_ to Warehouse,
which was merged on 13 April 2020.

The ``bdist_egg`` format
------------------------

The ``bdist_egg`` filetype identifies distributions in the

> `Egg format <https://packaging.python.org/en/latest/glossary/#term-Egg>`_. The

Reviewer 2: ```suggestion
:term:`Egg format <packaging:Egg>`. The
```

No need for a hardcoded link here—since we have `packaging.python.org` as an intersphinx target, you can crossreference the glossary term directly.

> Egg format was introduced by setuptools in 2004 and is roughly equivalent

Reviewer 5: Here and elsewhere, it's generally been written lower case in the past:

```suggestion
egg format was introduced by setuptools in 2004 and is roughly equivalent
```

e.g.

* https://setuptools.pypa.io/en/latest/deprecated/python_eggs.html
* http://peak.telecommunity.com/DevCenter/PythonEggs
* https://mail.python.org/archives/list/distutils-sig@python.org/thread/XON7O3BD2MWCMKGT5MPOCUHI6WKIPCST/#XON7O3BD2MWCMKGT5MPOCUHI6WKIPCST
* https://mail.python.org/pipermail/distutils-sig/2006-February/006026.html
* https://mail.python.org/pipermail/distutils-sig/2006-March/006123.html

Author: Thanks, I wasn't sure about that -- PyPUG uses "Egg" but most other sources seem to use "egg," so I'll standardize on the latter.

Author: https://github.com/python/peps/pull/3161/commits/fb85cc6e3d3038d3ea3e6d1f69c998473144d11e

in functionality to the

> `Wheel format <https://packaging.python.org/en/latest/glossary/#term-Wheel>`_

Reviewer 2: ```suggestion
:term:`Wheel format <packaging:Wheel>`
```

Likewise

> introduced and standardized by :pep:`427` in 2012.

Reviewer 2: ```suggestion
first introduced by :pep:`427` in 2012 as the standardized
:ref:`built distribution format <packaging:binary-distribution-format>`.
```

The canonical, up to date spec is the packaging.python.org document and the PEP is considered a historical change proposal, so we should update the text to reflect that.

Despite its longevity, the Egg format has had
`limited adoption on PyPI <https://github.com/pypi/warehouse/issues/10653>`_.
Some observations from that issue:

* As of December 2021, ``bdist_egg`` uploads account for less than 1% of all
  built distribution uploads;
* ``pip`` deprecated its ``--egg`` option in 2016;
* ``setuptools`` has considered Egg support deprecated since 2019;
* ``build`` only supports the ``sdist`` and ``bdist_wheel`` filetypes.

Given the above, this PEP proposes the removal of the ``bdist_egg`` format
under the same justifications presented in :pep:`527`, namely:

> 1. Egg distributions are of of limited use to the broader ecosystem and
>    therefore represent a non-reciprocal maintenance burden;
> 2. Continued support for Egg distributions represents a bandwidth and storage
>    cost to PyPI, as well as to any external mirrors of PyPI.

Reviewer 2: ```suggestion
* Egg distributions are of of limited use to the broader ecosystem and
  therefore represent a non-reciprocal maintenance burden;
* Continued support for Egg distributions represents a bandwidth and storage
  cost to PyPI, as well as to any external mirrors of PyPI.
```

This list doesn't appear to have an explicit, fixed order (as in a sequence of steps), so you might want to make it unordered instead.

This PEP *additionally* offers an argument for removal rooted in
standardization and duplication: the Egg format is not standardized by any
PEP or other community standard, and overlaps heavily with its standardized

> and well-supported alternative (Wheel).

Reviewer 5: Likewise for wheel (here and elsewhere):
```suggestion
and well-supported alternative (wheel).
```

https://pep-previews--3161.org.readthedocs.build/pep-0427/

Author: https://github.com/python/peps/pull/3161/commits/fb85cc6e3d3038d3ea3e6d1f69c998473144d11e

The ``.egg`` file extension
---------------------------

The ``.egg`` file extension is used exclusively for distributions of the
``bdist_egg`` format. As such, it serves no purpose in a scenario where
PyPI disables new distribution uploads for ``bdist_egg`` distributions.

Removal Process
===============

This PEP proposes a removal process based on :pep:`527`, but with the following

> modifications:

Reviewer 2: Given this is essentially the equivalent of the "specification", it ends up being not _that_ much shorter than the referenced section in PEP 527, and needing to both read that (rather dense) section and then apply the multiple deltas on top of that is not the easiest to understand, IMO could you just state the plan in full here in ordered-list or bullet point form (maybe retaining a brief mention of the deltas from PEP 527, perhaps in a subsection)?

Author: I've replaced this whole section with a single reference to PEP 527: [6d53cac](https://github.com/python/peps/pull/3161/commits/6d53cacec66c3a28ca8e53868c4c826f04d7220b)

(Let me know if that's okay -- otherwise, I can duplicate the whole process and make a small reference as you suggested.)

Author: Sounds good, doing!

Author: Done with 113abc18: I've more or less copied, with some reflow and edits to clarify that only `sdist` and `bdist_wheel` will be allowed after this PEP.

* During the deprecation period, *all* packages will
  `continue to be allowed <https://www.youtube.com/watch?v=XNyUALnj8V0>`_
  to upload ``bdist_egg`` distributions, not just those that have previously
  uploaded ``bdist_egg`` distributions.
* During the deprecation period, uploading a ``bdist_egg`` distribution to PyPI

>   will trigger an email to all maintainers of the containing package, rather

Reviewer 2: ```suggestion
  will trigger an email to all maintainers of the project, rather
```

Isn't it really a "project", not a "package" here? And "containing" seems rather confusing and unnecessary here, IMO.

------------------------

The ``bdist_egg`` filetype identifies distributions in the
:term:`egg format <packaging:Egg>`. The
egg format was introduced by setuptools in 2004 and is roughly equivalent
in functionality to the
:term:`wheel format <packaging:Wheel>`
first introduced by :pep:`427` in 2012 as the standardized
:ref:`built distribution format <packaging:binary-distribution-format>`.

Despite its longevity, the egg format has had
`limited adoption on PyPI <https://github.com/pypi/warehouse/issues/10653>`_.
Some observations from that issue:

* In the month of May 2023, ``bdist_egg`` uploads accounted for 0.2% of all
  distribution uploads to PyPI;
* ``pip`` deprecated its ``--egg`` option in 2016;
* ``setuptools`` has considered egg support deprecated since 2019;
* ``build`` only supports the ``sdist`` and ``bdist_wheel`` filetypes.

Given the above, this PEP proposes the removal of the ``bdist_egg`` format
under the same justifications presented in :pep:`527`, namely:

* egg distributions are of of limited use to the broader ecosystem and
  therefore represent a non-reciprocal maintenance burden;
* Continued support for egg distributions represents a bandwidth and storage

>   cost to PyPI, as well as to any external mirrors of PyPI.

Author: Agreed, thanks -- I'll emphasize that point rather than bandwidth/storage 🙂

------------------------

The ``bdist_egg`` filetype identifies distributions in the
:term:`egg format <packaging:Egg>`. The
egg format was introduced by setuptools in 2004 and is roughly equivalent
in functionality to the
:term:`wheel format <packaging:Wheel>`
first introduced by :pep:`427` in 2012 as the standardized

> :ref:`built distribution format <packaging:binary-distribution-format>`.

Author: Makes sense. This is stale in GitHub's review UI, so I'll manually include.

> :ref:`built distribution format <packaging:binary-distribution-format>`.

Author: 32ab4bae

Despite its longevity, the egg format has had
`limited adoption on PyPI <https://github.com/pypi/warehouse/issues/10653>`_.
Some observations from that issue:

* In the month of May 2023, ``bdist_egg`` uploads accounted for 0.2% of all
  distribution uploads to PyPI;
* ``pip`` deprecated its ``--egg`` option in 2016;
* ``setuptools`` has considered egg support deprecated since 2019;
* ``build`` only supports the ``sdist`` and ``bdist_wheel`` filetypes.

Given the above, this PEP proposes the removal of the ``bdist_egg`` format
under the same justifications presented in :pep:`527`, namely:

* Egg distributions are of of limited use to the broader ecosystem and
  therefore represent a non-reciprocal maintenance burden;

> * Having an additional bdist format is confusing to end users, who may

Author: Agreed, thank you!

Removal Process
===============

This PEP does **NOT** propose removing any existing files from PyPI, only
disallowing new ones from being uploaded.

PyPI will provide a deprecation period of one month. At the beginning
of the deprecation period, maintainers of projects that already have egg

> distributions will receive a one-time email informing them of the upcoming

Author: Oh, I was thinking this would be "uploaded one or more egg distributions *ever*", but maybe your edit is what @ewdurbin intended.

Either is fine by me, so punting to Ee for thoughts 🙂

> distributions will receive a one-time email informing them of the upcoming

Reviewer 2: Yup, I inferred the above from @ewdurbin 's [comments above](https://github.com/python/peps/pull/3161#discussion_r1220779290):

> Volumes are low enough I'm actually game to do the one-time email with a one month deprecation.

which was followed by

> > all 1022 projects and the number of eggs they've uploaded, in 2023: https://gist.github.com/ewdurbin/d7939458f6205069adeeb18c454c9331

(And the query in the linked Gist confirms that the list of projects are indeed those uploading an egg in 2023).

From this, I inferred that given that the referenced "volumes" and list of projects are those for 2023 (which would be at least ≈six months prior to the start of the deprecation period when this PEP is likely implemented, and following the release of Buildout versions supporting wheels on all active branches, and the EoL of those that support eggs), Ee was likely referring to this being the list of projects that would be notified, but I could definitely be interpreting that wrong. We'll see what they say.

> distributions will receive a one-time email informing them of the upcoming

Reviewer: Notify anyone who has uploaded in an egg in 2023.

> distributions will receive a one-time email informing them of the upcoming

Reviewer 2: Thanks @ewdurbin !

restrictions on egg distribution uploads.

During the deprecation period users will continue to be allowed to upload egg
distributions to new and existing projects. Uploading an egg distribution
during this period will result in an email to all maintainers of the project,
which will inform them of the upcoming restrictions on egg distribution
uploads.

After the deprecation period, support for uploading egg distributions will
cease to exist on PyPI.

Prior Art
---------

The removal process above is based on :pep:`527`'s removal process. However,
only the process above is authoritative for this PEP.

Backwards Compatibility
=======================

Limited Impact
--------------

As noted in the rationale section, this PEP is expected to have no impact

> on the overwhelming majority of PyPI users.

Reviewer 2: ```suggestion
on the overwhelming majority of PyPI users and projects, and there has been
`substantial community coordination <https://github.com/pypi/warehouse/issues/10653>`_
over the past 1 1/2 years to minimize the impact on the few last use cases.
```

You might want to add an explicit mention here of the substantial coordination work that's gone on with the community over the past 1.5 years to minimize the remaining impact.

Hosted Files
------------

This PEP does **NOT** propose the removal of any egg distributions that have
already been uploaded to PyPI. All previously uploaded egg distributions will
remain downloadable, ensuring that existing users will continue to be able
to download them.

Deprecation Period
------------------

This PEP uses the removal process documented above, which specifies
a deprecation period of 1 month for projects that have previously uploaded
egg distributions to PyPI.

After the end of the deprecation period, support for uploading new egg
distributions will cease to exist on PyPI.

Security Implications
=====================

This PEP does not identify any positive or negative security implications
associated with removing upload support for egg distributions.

How To Teach This
=================

As part of the removal process, PyPI will send emails to all maintainers of

> projects who have previously uploaded egg distributions.

Reviewer 2: ```suggestion
projects that have previously uploaded egg distributions in 2023.
```

* Clarify over what period
* Fix a word

Reviewer: ^
